src/libsyntax/parse/parser.rs

```diff
@@ -5194,98 +5161,155 @@ impl<'a> Parser<'a> {
+                TyKind::Path(None, path) => (TraitRef { path, ref_id: ty.id }, true),
+                _ => (TraitRef {
+                    path: ast::Path::from_ident(ty.span, keywords::Invalid.ident()),
+                    ref_id: ty.id
+                }, false),
             }
-            generics.where_clause = self.parse_where_clause()?;
-
-            self.expect(&token::OpenDelim(token::Brace))?;
-            let attrs = self.parse_inner_attributes()?;
+        };
 
-            let mut impl_items = vec![];
-            while !self.eat(&token::CloseDelim(token::Brace)) {
-                let mut at_end = false;
-                match self.parse_impl_item(&mut at_end) {
-                    Ok(item) => impl_items.push(item),
-                    Err(mut e) => {
-                        e.emit();
-                        if !at_end {
-                            self.recover_stmt_(SemiColonMode::Break, BlockMode::Break);
-                        }
-                    }
+        let item_kind = match ty_second {
+            SecondTy::Ty(ty_second) => {
+                // impl Trait for Type
+                if !has_for {
+                    self.span_err(missing_for_span, "missing `for` in a trait impl");
+                }
+                let (trait_ref, is_valid_trait) = ty_to_trait(ty_first);
+                if !is_valid_trait {
+                    self.span_err(trait_ref.path.span, "expected a trait, found type");
                 }
+                ItemKind::Impl(unsafety, polarity, defaultness,
+                               generics, Some(trait_ref), ty_second, impl_items)
             }
+            SecondTy::Dots => {
+                // impl Trait for ..
+                if !has_for {
+                    self.span_err(missing_for_span, "missing `for` in a trait impl");
+                }
+                let (trait_ref, is_valid_trait) = ty_to_trait(ty_first);
+                if !is_valid_trait {
+                    self.span_err(trait_ref.path.span, "expected a trait, found type");
+                }
+                if generics.is_parameterized() {
+                    self.span_err(generics.span, "auto impls cannot have generic parameters");
+                }
+                if !generics.where_clause.predicates.is_empty() {
+                    self.span_err(generics.where_clause.span,
+                                  "auto impls cannot have predicates in `where` clauses");
+                }
+                if polarity == ast::ImplPolarity::Negative {
+                    self.span_err(polarity_span, "auto impls cannot be negative");
+                }
+                if defaultness == Defaultness::Default {
+                    self.span_err(defaultness_span, "auto impls cannot be default");
+                }
+                if !impl_items.is_empty() {
+                    self.span_err(impl_items[0].span, "auto impls cannot have impl items")
+                }
```

Comment:
I didn't add new tests for all this stuff because impls for `..` are going to be removed pretty soon.

Reply:
I think we can even do it now, if we do some more `#[cfg(stage0)]` stuff.